python/ray/reporter.py

```diff
@@ -0,0 +1,208 @@
+    def get_cpu_percent():
+        return psutil.cpu_percent()
+
+    @staticmethod
+    def get_boot_time():
+        return psutil.boot_time()
+
+    @staticmethod
+    def get_network_stats():
+        now = datetime.datetime.utcnow()
+        ifaces = [
+            v
+            for k, v in psutil.net_io_counters(pernic=True).items()
+            if k[0] == "e"
+        ]
+
+        sent = sum((iface.bytes_sent for iface in ifaces))
+        recv = sum((iface.bytes_recv for iface in ifaces))
+        return sent, recv
+
+    @staticmethod
+    def get_mem_usage():
+        vm = psutil.virtual_memory()
+        return vm.total, vm.available, vm.percent
+
+    @staticmethod
+    def get_disk_usage():
+        return {
+            x: psutil.disk_usage(x)
+            for x in ["/", "/tmp"]
+        }
+
+    @staticmethod
+    def get_workers():
+        return [
+            x.as_dict(attrs=["pid", "create_time", "cpu_times", "name", "memory_full_info"])
+            for x in psutil.process_iter()
+            if running_worker(x.name())
+        ]
+
+    def get_load_avg(self):
+        load = os.getloadavg()
+        per_cpu_load = tuple(
+            ( round(x/self.cpu_counts[0], 2) for x in load )
+        )
+        return load, per_cpu_load
+
+    def get_all_stats(self):
+        now = to_posix_time(datetime.datetime.utcnow())
+        network_stats = self.get_network_stats()
+
+        self.network_stats_hist.append((now, network_stats))
+        self.network_stats_hist = self.network_stats_hist[-7:]
+        then, prev_network_stats = self.network_stats_hist[0]
+        netstats = (
+            (network_stats[0]-prev_network_stats[0])/(now - then),
+            (network_stats[1]-prev_network_stats[1])/(now - then)
+        )
+
+        return {
+            "now": now,
+            "hostname": self.hostname,
+            "ip": self.ip_addr,
+            "cpu": self.get_cpu_percent(),
+            "cpus": self.cpu_counts,
+            "mem": self.get_mem_usage(),
+            "workers": self.get_workers(),
+            "boot_time": self.get_boot_time(),
+            "load_avg": self.get_load_avg(),
+            "disk": self.get_disk_usage(),
+            "net": netstats,
+        }
+
+    def perform_iteration(self):
+        """Get any changes to the log files and push updates to Redis."""
+        stats = self.get_all_stats()
+
+        self.redis_client.publish(
+            self.redis_key,
+            simplejson.dumps(stats, namedtuple_as_object=True),
+        )
+
+    def run(self):
+        """Run the reporter."""
+        while True:
+            try:
+                self.perform_iteration()
+            except Exception:
+                traceback.print_exc()
+                pass
+
+            time.sleep(ray_constants.REPORTER_UPDATE_INTERVAL_MS/1000)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description=("Parse Redis server for the "
+                     "reporter to connect to."))
```

Comment:
--http-port? We can do something similar to the "old" web UI, and incrementally try higher ports until a free one is found.

Reply:
Yep.

Reply:
Done.